Add EV power and current streams to `EVChargerPool`

shsms opened the pull request:
These formulas were getting generated from the component graph in the logical meter.  

Those have now been removed,  and the formulas are now generated for the ev chargers that are part of the ev charger pool.

Additional methods have been added to fetch the power and current values for individual ev chargers.

shsms committed:
Move FormulaEngine and formula generators into a separate package

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Extract EVChargerPool's StateTracker feature into a separate module

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Extract formula ownership/reuse code into a separate class

... called `FormulaEnginePool`.  This code was previously part of the
`LogicalMeter`, but is required in the other *Pools as well.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Extract formula engine test utils into a separate package

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move formula composition tests out of `test_logical_meter.py`

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update `MockMicrogrid.start` to return a channel instead of a sender

So that channel senders don't get reused in tests.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Move ev charger power and current formulas to EVChargerPool

... from the logical meter

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add supporting for configuring formula generators

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Update EV formula generators to use component_ids from the ev pool

... instead of from the component graph.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms committed:
Add functions to fetch current and power for individual ev chargers

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,173 @@
+            *[api_client.ev_charger_data(cid) for cid in self._component_ids]
+        )
+
+        latest_messages: list[EVChargerData] = await asyncio.gather(
+            *[stream.receive() for stream in streams]
+        )
```
In batteries/inverters component might be visible but won't send any data. 
I am not sure how it works in ev chargers, but you might wait forever here.
You could wait with some timeout, here.

shsms replied:
True,  I guess I also have to add code to switch state to error if there's no data from any ev charger for a certain time.  Will add both.

shsms replied:
I would like to try this in the ev charging actor first and make improvements later.  I think that would give a lot more clarity and time to think.

The EVChargerPool that was already merged has these same issues as well,  so we are not removing an existing feature.

Instead,  I'll add a note in the docstring that the `states` method is not ready for production use and create an issue to do this in a subsequent PR.

shsms replied:
I've updated the PR such that the `StateTracker` is unused.  I'm modifying it in the next PR to handle the above cases in alternative ways discussed during the actor meeting.

ela-kotulska-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,108 @@
+        _GenericEngine,
+        _GenericFormulaReceiver,
+    )
+    from ._formula_generators import FormulaGenerator
```
Could you add comment why use that `if TYPE_CHECKING`?

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,116 @@
+        self._namespace = namespace
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
+        self._engines: dict[str, "FormulaReceiver|FormulaReceiver3Phase"] = {}
```
why some attributes of the class add type hints and some other don't? I'm asking here as I didn't want to interrupt you during the meeting/presentation

daniel-zullo-frequenz replied:
maybe because the attributes set from the input parameters don't need it?

shsms replied:
yup,  in the case of `self._engines`,  it can't mypy can't infer the type from the assigned value,  because it is an empty dict.  For the others,  it infers the type from the type of the assigned value.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,173 @@
+        self._component_ids = component_ids
+        self._channel = Broadcast[EVChargerPoolStates](
+            "EVCharger States", resend_latest=True
+        )
```
related to https://github.com/frequenz-floss/frequenz-sdk-python/pull/201/files#r1108673277, the type hint might be missing here though

shsms replied:
Here the typehint is inferred by the convenience thing in python type hints:

``` python
self._channel: Broadcast[int] = Broadcast("name")

## is the same as:

self._channel = Broadcast[int]("name")
```

daniel-zullo-frequenz replied:
Got it, thanks!

shsms committed:
Remove `states` method for streaming ev charger state changes

This will be replaced by a method that streams states along with power
and current metrics for individual ev chargers.

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

shsms commented:
I've removed the `states` method from the EVChargerPool, in the last commit.

The `StateTracker` implementation is no longer publicly exposed, and will need some changes to work with the new approach for exposing component states.  I will make those in a new PR.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/__init__.py:
```diff
@@ -0,0 +1,25 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
2023?

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,117 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
ditto

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,117 @@
+            channel_registry: A channel registry instance shared with the resampling
+                actor.
+            resampler_subscription_sender: A sender for sending metric requests to the
+                resampling actor.
```
I think the variable name is too long, can `subscription` be removed as it is not even mentioned in the documentation?

shsms replied:
This is the same name across all our data pipeline actors and logical_meter.  So it is a broader topic.  

For this PR,  it is probably more sensible to be consistent with existing code, but I agree and it would be good to track and change everywhere in one go.  Maybe you can bring this up in the next sdk meeting.

daniel-zullo-frequenz replied:
makes sense. Yes, we can discuss it in the meeting

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,117 @@
+class FormulaEnginePool:
+    """Creates and owns formula engines from string formulas, or formula generators.
+
+    If an engine already exist with a given name, it is reused instead.
```
```suggestion
    If an engine already exists with a given name, it is reused instead.
```

shsms replied:
fixed.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,117 @@
+            formula: formula to execute.
+            component_metric_id: The metric ID to use when fetching receivers from the
+                resampling actor.
+            nones_are_zeros: Whether to treat None values from the stream as 0s.  If
+                False, the returned value will be a None.
```
I wonder why docstrings does not catch double-spaces:

```suggestion
            nones_are_zeros: Whether to treat None values from the stream as 0s. If
                False, the returned value will be a None.
```

shsms replied:
This was not a mistake.  I went to a typing school to learn touch typing on a type writer,  so I double space my sentences. :-)

Also,  emacs, vim, etc. use double spaces to differentiate between sentence breaks and abbreviations (like `etc.` or `eg.`), when doing sentence level operations, so I like to use them too.

daniel-zullo-frequenz replied:
alright, then let's keep it :)

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_engine_pool.py:
```diff
@@ -0,0 +1,117 @@
+    async def from_generator(
+        self,
+        channel_key: str,
+        generator: "Type[FormulaGenerator[_GenericEngine]]",
```
Is this type hint definition due to the circular import? I personally find it hacky but I guess there is not other workaround. Ideally we shouldn't have circular imports dependencies

shsms replied:
This is a Python problem,  it needs hacks to get around its circular imports problem.  And this hack is the recommended way to deal with it.

Quoting from: https://docs.python.org/3/library/typing.html#typing.TYPE_CHECKING

> The first type annotation must be enclosed in quotes, making it a “forward reference”, to hide the expensive_mod reference from the interpreter runtime. Type annotations for local variables are not evaluated, so the second annotation does not need to be enclosed in quotes.

daniel-zullo-frequenz replied:
Oh I see, thanks for pointing that out

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_ev_charger_current_formula.py:
```diff
@@ -18,21 +18,16 @@ class EVChargerCurrentFormula(FormulaGenerator):
+        if not component_ids:
             logger.warning(
-                "Unable to find any EV Chargers in the component graph. "
+                "No ev charger component IDs specified. "
```
```suggestion
                "No EV charger component IDs specified. "
```

shsms replied:
done (the capitalization issues are fixed in a separate commit)

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_ev_charger_power_formula.py:
```diff
@@ -17,22 +16,17 @@ class EVChargerPowerFormula(FormulaGenerator):
 
     async def generate(self) -> FormulaEngine:
-        """Generate a formula for calculating total EV power from the component graph.
+        """Generate a formula for calculating total ev power for given component ids.
```
```suggestion
        """Generate a formula for calculating total EV power for given component ids.
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_ev_charger_power_formula.py:
```diff
@@ -17,22 +16,17 @@ class EVChargerPowerFormula(FormulaGenerator):
+        if not component_ids:
             logger.warning(
-                "Unable to find any EV Chargers in the component graph. "
+                "No ev charger component IDs specified. "
```
```suggestion
                "No EV charger component IDs specified. "
```

shsms replied:
done

mathias-baumann-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,172 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
2023

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -5,147 +5,51 @@
         """
-        self._component_ids = set()
+        self._channel_registry = channel_registry
+        self._resampler_subscription_sender = resampler_subscription_sender
```
This attribute name is too long

shsms replied:
I like to keep the attribute name the same as the parameter,  but related to: https://github.com/frequenz-floss/frequenz-sdk-python/pull/201#discussion_r1116812353

so it can be changed separately, later.

daniel-zullo-frequenz replied:
This will be kept as it is

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
-            *[api_client.ev_charger_data(cid) for cid in self._component_ids]
+
+    async def total_current(self) -> FormulaReceiver3Phase:
+        """Fetch the total current for the ev chargers in the pool.
```
```suggestion
        """Fetch the total current for the EV chargers in the pool.
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
+    async def total_current(self) -> FormulaReceiver3Phase:
+        """Fetch the total current for the ev chargers in the pool.
+
+        If a formula engine to calculate ev charger current is not already
+        running, it will be started.  Else, we'll just get a new receiver to the
+        already existing data stream.
```
```suggestion
        If a formula engine to calculate EV charger current is not already
        running, it will be started. Else, it'll just get a new receiver to the
        already existing data stream.
```

daniel-zullo-frequenz replied:
please discard the double-space suggestion but the other ones still might apply

shsms replied:
reworded the "It'll get a" to "It'll return a" in a separate commit.

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
-        latest_messages: list[EVChargerData] = await asyncio.gather(
-            *[stream.receive() for stream in streams]
+    async def total_power(self) -> FormulaReceiver:
+        """Fetch the total power for the ev chargers in the pool.
```
```suggestion
        """Fetch the total power for the EV chargers in the pool.
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
+        """Fetch the total power for the ev chargers in the pool.
+
+        If a formula engine to calculate EV charger power is not already
+        running, it will be started. Else, we'll just get a new receiver
```
```suggestion
        running, it will be started. Else, it'll just get a new receiver
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
+        """Fetch the 3-phase current for the given ev_charger id.
+
+        Args:
+            component_id: id of the ev charger to stream current values for.
```
```suggestion
            component_id: id of the EV charger to stream current values for.
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_ev_charger_pool.py:
```diff
@@ -156,47 +60,88 @@ def __init__(
+        """Fetch the power for the given ev_charger id.
+
+        Args:
+            component_id: id of the ev charger to stream power values for.
```
```suggestion
            component_id: id of the EV charger to stream power values for.
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,172 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
```suggestion
# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```

shsms replied:
done

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,172 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""State tracking for ev charger pools."""
```
General suggestion, please capitalize `ev` if it is used in the documentation and it isn't part of a variable/attribute name

daniel-zullo-frequenz replied:
please check update all occurrences from ` ev ` to ` EV ` if this suggestion is taken

shsms replied:
thanks,  I've fixed all occurrences.

daniel-zullo-frequenz replied:
Thanks!

shsms committed:
Fix capitalization for "EV Chargers" in docstrings and log messages

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

daniel-zullo-frequenz commented on src/frequenz/sdk/timeseries/ev_charger_pool/_state_tracker.py:
```diff
@@ -0,0 +1,172 @@
+
+        Returns:
+            A tuple with the component ID of an ev charger that just had a state
+                change, and its new state.
```
it can also return `None` and this is not explicitly mentioned in the documentation

shsms replied:
fixed.

daniel-zullo-frequenz commented on tests/timeseries/_formula_engine/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
```suggestion
# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
```

shsms replied:
fixed.

daniel-zullo-frequenz commented on tests/timeseries/_formula_engine/test_formula_composition.py:
```diff
@@ -0,0 +1,180 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
```
General suggestion: New files should set 2023 as a year

daniel-zullo-frequenz replied:
Please update the year in all new files added if this suggestion is taken

shsms replied:
done.

daniel-zullo-frequenz commented on tests/timeseries/_formula_engine/test_formula_composition.py:
```diff
@@ -0,0 +1,180 @@
+
+            assert net_amps.value_p1 == grid_amps.value_p1 - ev_amps.value_p1
+            assert net_amps.value_p2 == grid_amps.value_p2 - ev_amps.value_p2
+            assert net_amps.value_p3 == grid_amps.value_p3 - ev_amps.value_p3
```
if these are `float` types values we shouldn't then use `==` operator for comparison. I guess this is fine for tests ?

shsms replied:
yes,  these tests pass,  and we're really just moving code around here.

daniel-zullo-frequenz replied:
right, resolving...

shsms committed:
Improve documentation for generated formula stream methods

Signed-off-by: Sahas Subramanian <sahas.subramanian@proton.me>

daniel-zullo-frequenz reviewed (commented):
I have only minor suggestions to check for and LGTM otherwise
